### PR TITLE
Added support for advanced label filters when locking

### PR DESCRIPTION
**Pull request by gaspardpetit**
Finding resources based on labels is nice, but the current implementation is very limiting. We would need to be able to filter out some resources.

Typically when implementing filters, my experience is that a trio of all/any/none is sufficient for most use cases, so instead of just adding exclusion I added all three.

- `anyOfLabels` will consider resources if they have at least one matching label
- `allOfLabels` will consider resources if *all* labels match
- `noneOfLabels` will consider resources unless any of their label matches

The previously used `label` is still supported and considered a "simple" case of finding a resource using a single label. In implementation, this value is merged into the `allOfLabels` down the road.

The three new filters are placed under an "Advanced" group in the jelly.

I hope this is useful to others as well. Happy to make any changes necessary to get this PR in.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkins/plugins/lockableresources/queue/LockableResourcesQueueTaskDispatcher.java [hunk 1]
```diff
@@ -197,10 +207,18 @@ public BecauseResourcesQueueFailed(
     @Override
     public String getShortDescription() {
       // TODO: Just a copy-paste from BecauseResourcesLocked, seems strange
```
Comment (jimklimov) [inline src/main/java/org/jenkins/plugins/lockableresources/queue/LockableResourcesQueueTaskDispatcher.java:209]: Probably this TODO no longer applies? :)

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java [hunk 2]
```diff
@@ -164,12 +166,8 @@ public int getFreeResourceAmount(String label) {
     return free;
   }
 
-  public List<LockableResource> getResourcesWithLabel(String label, Map<String, Object> params) {
```
Comment (jimklimov) [inline src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java:167]: Wondering if it is fair to drop the earlier `public` method, there may be pipelines using it (directly calling LRM)... To err on safe side, I'd keep it, just maybe mark as `@Deprecated` or similar.

**Review by jimklimov — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java (hunk 2 above)
Comment (gaspardpetit) [reply src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java:167]: I don't mind keeping it around, the `params` are actually unused - I'd be tempted to remove it but if we're going to remove the whole thing in a few releases, I guess we might as well keep it?

**Comment by gaspardpetit** [comment]
Thanks for the review, I added back the `getResourcesWithLabel` with deprecation notice. I have been using this change in our production build Jenkins for over a month now and it's working great. In our case, we allow labeling some resources as 'offline' (using https://github.com/jenkinsci/lockable-resources-plugin/pull/305) and then filter on `lock(extra:[quantity:1, allOfLabel:'console', noneOfLabel:'offline'])`.  Happy to make any other changes that can get this into the main branch.

**Comment by chickenkiller** [comment]
I love this addition ! Any chance to see it merged soon?

**Comment by DonnieKim411** [comment]
Any update on this? this feature will be super useful for our use case

**Comment by mPokornyETM** [comment]
will be handled in #428

**Comment by monger39** [comment]
would this feature also support allocating a resource to a specific node (eg by name reference) , then checking as one of the conditions that such node is online? Eg image 6 nodes each having a printer, for which some printer_NN resource is defined with label 'printer'. How to lock a printer resource from an online node?

**Comment by mPokornyETM** [comment]
> would this feature also support allocating a resource to a specific node (eg by name reference) , then checking as one of the conditions that such node is online? Eg image 6 nodes each having a printer, for which some printer_NN resource is defined with label 'printer'. How to lock a printer resource from an online node?

There are few similar PRs and no one is completed (for some historical reason). At the moment it is hard to merge master in this branches.
Therefore will be this PR probably closed and handled in #428. 
To your question about node filters.
I have also created #341 which is similar idea.
When you need to use some special (more sophisticated) filter just now, you can use a Groovy expression. When it does not match your requirements, please create a new enhancement issue here in GitHub. It will be also helpful to provide a small example, how you want to use it in your pipeline?

PS: Strange for some reason is the Groovy expression visible only in free-style jobs. This shall be fixed. Please help me and create new issue here in GitHub (thx)
image

PS2: currently I am thinking about new global-lockable-resources-shared-library which extends core functionality in groovy. Because of historical reason, are many things no more possible here (without big effort, by refactoring this plugin). This can be loaded by you, in your Jenkins instance and used for free. BUT it need time and we shall specified which content shall be done there. Currently nobody will contribute (there)[https://github.com/jenkinsci/lockable-resources-plugin/discussions/445]. Maybe you will find some time and help me (us). It does not mean to contribute on coding. Describe an idea in new issues helps too ;-)

**Comment by monger39** [comment]
@mPokornyETM : I have created a new issue https://github.com/jenkinsci/lockable-resources-plugin/issues/455 describing my idea of the use-case. I would be even willing to help on the coding part, but am an absolute newbie on plugin development (and Java ;-) )

**Comment by mPokornyETM** [comment]
> @mPokornyETM : I have created a new issue #455 describing my idea of the use-case. I would be even willing to help on the coding part, but am an absolute newbie on plugin development (and Java ;-) )

Any help is welcome. Contributing is not only about Java.
Documentation, testing, PR-reviews, bug-management, localizations, bring new ideas and many more.
Please join our chat and we can discuss any thing you want ;-)
https://gitter.im/jenkinsci/lockable-resources

**Pull request closed without merging**